lib/patches/file_downloaders/s3_downloader.rb

```diff
@@ -0,0 +1,35 @@
+require 'aws-sdk-s3'
+
+module Patches
+  module FileDownloaders
+    class S3Downloader
+      def initialize(bucket_name:, aws_config:)
+        @bucket_name = bucket_name
+        @aws_config = aws_config
+      end
+
+      def download(filename, destination)
+        update_aws_config
+        get_remote_file(filename, destination)
+        destination
+      end
+
+      private
+
+      def update_aws_config
+        Aws.config.update({
+          region: @aws_config[:region],
+          credentials: Aws::Credentials.new(@aws_config[:access_key_id], @aws_config[:secret_access_key])
```

Comment:
I'm not entirely clear whether this requires the use of the access and secret keys or would work without them, e.g. if an IAM instance role is available.
https://github.com/jobready/esherhouse/blob/develop/app/services/s3_downloader.rb#L20
shows one way of using the IAM role.

The AWS libraries use the same credential preference as the CLI tool.
https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#id1
In fact we can rely on the defaults of the IAM Role being available or the right env vars being set to manage access.

Reply:
Not sure what IAM is and how it works. The esherhouse link doesn't really show how to use but only how to instantiate a resource with no credentials. 

Could you please advise?

Reply:
All the AWS libraries use standard conventions to access credentials.
https://docs.aws.amazon.com/sdk-for-ruby/v3/api/index.html#Configuration
talks about it a bit.
IAM == Identity and Access Management
IAM Role == A role given to an EC2 instance that allows it to obtain auto-rotating temporary credentials e.g. (never configure creds again). This is much better than passing creds in.

EH doesn't need to specify credentials because it can auto obtain them from the IAM role. The same can be done with access keys if the right env vars are used. e.g. `AWS_ACCESS_KEY`. This seems like a worthy convention to follow.